### PR TITLE
Rewrite rack and rack load

### PR DESCRIPTION
**Pull request by tjadams**
This is the first pass through the rewrite.
I still need to write some tests and verify that this works.

### CODE REVIEW

File: src/main/scala/Rack/RackActor.scala [hunk 1]
```diff
@@ -1,39 +1,82 @@
...
 import akka.actor.{Actor, ActorRef}
+import messageTypes.{PowerRestored, PowerLoss, EventType}
 
-/**
- * Created by jakewang on 10/12/15.
- */
 class RackActor (breakerRef:ActorRef, rackName:String) extends Actor {
 
   val name = "[Rack: " + rackName + "]"
   val id = UUID.random
```
Comment (swmyuen) [inline src/main/scala/Rack/RackActor.scala]: use UUID class i made

File: src/main/scala/Rack/RackActor.scala [hunk 2]
```diff
@@ -1,39 +1,82 @@
...
-  breakerHandler ! "register"
+  breakerRef ! "register" // have this rack registered with upstream
+
+//  case class requestHandler(request:DcsimRequest)
+//  case class eventHandler(eventType: EventType)
+//  case class register(deviceType:DeviceType)
 
   def receive = {
-    case _ =>
+//    case requestHandler(request) => requestHandler(request, sender)
```
Comment (swmyuen) [inline src/main/scala/Rack/RackActor.scala]: remove unused lines

File: src/main/scala/Rack/RackActor.scala [hunk 3]
```diff
@@ -1,39 +1,82 @@
...
+    case ("request", request:DcsimRequest) => requestHandler(request, sender)
+//    case eventHandler(event) => eventHandler(event, sender)
+    case ("event", event:EventType) => eventHandler(event, sender)
+//    case register(device) => registerDevice(device, sender)
+    case ("register", deviceType:DeviceType, device:ActorRef) => registerDevice(deviceType, device, sender)
```
Comment (swmyuen) [inline src/main/scala/Rack/RackActor.scala]: These case statements should all be enums and it should not include the strings

File: src/main/scala/Rack/RackActor.scala [hunk 4]
```diff
@@ -1,39 +1,82 @@
...
 
-  def requestHandler() = {
+  def requestHandler(request:DcsimRequest, sender:ActorRef) = request.requestType match {
+    case LoadChanged => {
+      upsRef ! ("request", request)  // receive a loadChanged request from RackLoad and send that to the UPS
+      // TODO write a test to make sure this goes through the UPSActor receive method into the UPSActor requestHandler
+    }
+
+    case PowerAllocation => {
+      breakerRef ! ("request", request) // receive a PowerAllocation request from UPS and send that request to the breaker
```
Comment (swmyuen) [inline src/main/scala/Rack/RackActor.scala]: Use enums and not strings

File: src/main/scala/Rack/RackActor.scala [hunk 5]
```diff
@@ -1,39 +1,82 @@
...
+      // TODO write a test to make sure this goes through the BreakerActor receive method into the BreakerActor requestHandler
+    }
 
+    case _ => log.error("Unexpected request received by "+name +" from sender: "+sender)
   }
 
-  def eventHandler() = {
+  def eventHandler(event:EventType, sender:ActorRef) = event match {
+    case PowerLoss => {
+      upsRef ! ("event", event)  // receive a PowerLoss event from Breaker and send that to the UPS
```
Comment (swmyuen) [inline src/main/scala/Rack/RackActor.scala]: Use enums and not strings

File: src/main/scala/Rack/RackActor.scala [hunk 6]
```diff
@@ -1,39 +1,82 @@
...
+    }
 
-  }
+    case PowerRestored => {
+      upsRef ! ("event", event) // receive a PowerRestored event from Breaker and send that to the UPS
+    }
```
Comment (swmyuen) [inline src/main/scala/Rack/RackActor.scala]: Use enums and not strings

File: src/main/scala/Rack/RackActor.scala [hunk 7]
```diff
@@ -1,39 +1,82 @@
...
-  def registerDevice(deviceType:String, sender : ActorRef) = deviceType match {
-    case "ups" =>
-    case "rackLoad" =>
+    case _ => {
+      log.error("Bad event received in RackActor: "+event)
+      sender ! "bad event"
+    }
   }
 
+  def registerDevice(deviceType:DeviceType, deviceRef : ActorRef, sender : ActorRef) = deviceType match {
```
Comment (swmyuen) [inline src/main/scala/Rack/RackActor.scala]: Are all of these arguments needed?

File: src/main/scala/Rack/RackActor.scala [hunk 8]
```diff
@@ -1,39 +1,82 @@
...
+    case Ups => {
+      upsRef = deviceRef
+      breakerRef ! ("request", UpsRegistration)
```
Comment (swmyuen) [inline src/main/scala/Rack/RackActor.scala]: Use enums and not strings

File: src/main/scala/Rack/RackActor.scala [hunk 9]
```diff
@@ -1,39 +1,82 @@
...
+      // TODO how do I handle the errors here as in Node.js...
+    }
+
+    case RackLoad => {
+      rackLoadRef = deviceRef
+      breakerRef ! ("request", RackLoadRegistration) // TODO make sure that Jacob's code for requestHandler RackLoadRegistration says upstreamBreaker ! ("request", request)
```
Comment (swmyuen) [inline src/main/scala/Rack/RackActor.scala]: Use enums and not strings

File: src/main/scala/Rack/RackLoadActor.scala [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+import akka.actor.{Actor, ActorRef}
+import akka.event.Logging
+import api.{DcsimRequest, LoadChanged}
+import messageTypes._
+import io.jvm.uuid._
+
+class RackLoadActor (rackRef : ActorRef, rackName : String) extends Actor {
+  val name = "[Rack Load: " + rackName + "]"
+  val id = UUID.random
```
Comment (swmyuen) [inline src/main/scala/Rack/RackLoadActor.scala]: Use UUID method i made

File: src/main/scala/Rack/RackLoadActor.scala [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  // register the load on rack
+  rackRef ! "register"
+
+  //  case class eventHandler(eventType: EventType)
+
+  def receive = {
+//    case requestHandler(request) => requestHandler(request)
+//    case ("request", request:DcsimRequest) => requestHandler(request)
+    //    case eventHandler(event) => eventHandler(event)
+    case ("event", event:EventType) => eventHandler(event)
```
Comment (swmyuen) [inline src/main/scala/Rack/RackLoadActor.scala]: Use enums and not strings

File: src/main/scala/Rack/RackLoadActor.scala [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    case _ => log.error("Unexpected message received by "+name)
+  }
+
+  def setLoad(load:Integer) = {
+    currentLoad = load
+    var request : DcsimRequest = new DcsimRequest(LoadChanged, id, currentLoad, self)
```
Comment (swmyuen) [inline src/main/scala/Rack/RackLoadActor.scala]: Is there a reason that we need all of these arguments?

File: src/main/scala/Rack/RackLoadActor.scala [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    rackRef ! ("request", request)
```
Comment (swmyuen) [inline src/main/scala/Rack/RackLoadActor.scala]: Use enums and not strings

File: src/main/scala/Rack/RackLoadActor.scala [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+//
+//    case GetRackLoad => {
+//
+//    }
+//
+//    case _ => log.error("Unexpected request received by "+name +" from sender: "+sender)
+//  }
+
+  def eventHandler(event : EventType) = event match {
+    case PowerLoss => { // lost all power from upstream breaker
```
Comment (swmyuen) [inline src/main/scala/Rack/RackLoadActor.scala]: Comment should go on the line above case PowerLoss =>

File: src/main/scala/api/RequestType.scala [hunk 15]
```diff
@@ -8,6 +8,8 @@ sealed trait RequestType {def Type : String}
 case object UpsRegistration extends RequestType {val Type = "upsRegistration"}
 case object RackLoadRegistration extends RequestType {val Type = "rackLoadRegistration"}
 case object PowerAllocation extends RequestType {val Type = "powerAllocation"}
-
 case object LoadChanged extends RequestType {val Type = "loadChanged"}
 case object PeakReset extends RequestType {val Type = "peakReset"}
+//case object GetRackLoad extends RequestType {val Type = "getRackLoad"}
```
Comment (swmyuen) [inline src/main/scala/api/RequestType.scala]: Remove commented stuff if not needed

File: src/main/scala/Rack/RackActor.scala [hunk 16]
```diff
@@ -1,39 +1,82 @@
...
-  var ups : UpsActor = null
-  val breakerHandler : ActorRef = breakerRef
+  val log = Logging(context.system, this)
+  log.info("Creating "+name+" ("+id+")")
 
-  breakerHandler ! "register"
+  breakerRef ! "register" // have this rack registered with upstream
+
+//  case class requestHandler(request:DcsimRequest)
+//  case class eventHandler(eventType: EventType)
```
Comment (swmyuen) [inline src/main/scala/Rack/RackActor.scala]: remove unused lines

File: src/main/scala/Rack/RackActor.scala [hunk 17]
```diff
@@ -1,39 +1,82 @@
...
+  var upsRef : ActorRef = null
 
-  var rackLoad : RackLoadActor = null
-  var ups : UpsActor = null
-  val breakerHandler : ActorRef = breakerRef
+  val log = Logging(context.system, this)
+  log.info("Creating "+name+" ("+id+")")
 
-  breakerHandler ! "register"
+  breakerRef ! "register" // have this rack registered with upstream
```
Comment (swmyuen) [inline src/main/scala/Rack/RackActor.scala]: remove unused lines

**Comment by swmyuen** [comment]
Main things:

Use enums not strings 
Remove un-used commented out lines

File: src/main/scala/Rack/RackActor.scala (hunk 7 above)
Comment (tjadams) [inline src/main/scala/Rack/RackActor.scala]: the sender is not needed but the deviceRef and deviceType are

File: src/main/scala/Rack/RackLoadActor.scala (hunk 12 above)
Comment (tjadams) [inline src/main/scala/Rack/RackLoadActor.scala]: Yeah. They're used in the BreakerActor

**Comment by tjadams** [comment]
@elkanwang , want to take a look?

File: src/main/scala/Rack/RackActor.scala [hunk 18]
```diff
@@ -1,39 +1,81 @@
...
+import api._
 import akka.actor.{Actor, ActorRef}
+import messageTypes.{PowerRestored, PowerLoss, EventType}
 
-/**
- * Created by jakewang on 10/12/15.
- */
 class RackActor (breakerRef:ActorRef, rackName:String) extends Actor {
 
   val name = "[Rack: " + rackName + "]"
```
Comment (elkanwang) [inline src/main/scala/Rack/RackActor.scala]: We are not using the name anymore. We use actorRef instead

File: src/main/scala/Rack/RackLoadActor.scala [hunk 19]
```diff
@@ -0,0 +1,50 @@
+package Rack
+
+import Utils.UpsUtilities
+import akka.actor.{Actor, ActorRef}
+import akka.event.Logging
+import api.{RackLoad, DeviceType, DcsimRequest, LoadChanged}
+import messageTypes._
+
+class RackLoadActor (rackRef : ActorRef, rackName : String) extends Actor {
+  val name = "[Rack Load: " + rackName + "]"
```
Comment (elkanwang) [inline src/main/scala/Rack/RackLoadActor.scala]: same thing

File: src/main/scala/Rack/RackLoadActor.scala [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+  val log = Logging(context.system, this)
+  log.info("Creating", name, "(", id, ")")
+
+  case class Event(eventType: EventType)
+  case class Request(request: DcsimRequest)
+  case class RegisterWithRef(deviceType:DeviceType, device:ActorRef)
+
+  // register the load on it's rack
+  rackRef ! RegisterWithRef(RackLoad, self)
```
Comment (elkanwang) [inline src/main/scala/Rack/RackLoadActor.scala]: self is sent to the target automatically

Comment (swmyuen) [inline src/main/scala/Rack/RackLoadActor.scala]: +1

Comment (tjadams) [inline src/main/scala/Rack/RackLoadActor.scala]: it's `sender()` right?

**Comment by tjadams** [comment]
@elkanwang @swmyuen I updated the PR

**Pull request merged**